File: 114-flatten-binary-tree-to-linked-list.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var pre *TreeNode
func flatten(root *TreeNode) {
	pre = nil
	helper(root)
}

func helper(nd *TreeNode) {
	if nd == nil {
		return
	}
	fmt.Println(nd.Val)
	if pre == nil {
		pre = nd
	} else {
		pre.Right = nd
		pre = nd
	}
	left := nd.Left
	right := nd.Right
	nd.Left = nil
	nd.Right = nil
	helper(left)
	helper(right)
}

File: bwu/binary_tree/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if root is None:
            return
        
        dummy = TreeNode(0)
        
        def preOrder(prev, root):
            stack = [root]
            while stack:
                prev.right = stack[-1]
                node = stack.pop()
                prev = node
                if node.right:
                    stack.append(node.right)
                if node.left:
                    stack.append(node.left)
                    node.left = None
                    
        preOrder(dummy, root)
